### PR TITLE
initial implementation of task priorities

### PR DESCRIPTION
**Pull request by andre-merzky**

### CODE REVIEW

**Comment by andre-merzky** [comment]
I would not mind getting this merged - the test case passes, tests pass, it fulfills the requested semantics.  The use case tests seems to take a bit longer, but I don't think the PR should be held up by that anymore.  @mturilli ?

File: src/radical/pilot/agent/scheduler/base.py [hunk 1]
```diff
@@ -892,61 +907,70 @@ def _schedule_incoming(self):
```
Comment (mtitov) [inline src/radical/pilot/agent/scheduler/base.py:907]: > `self.slot_status("before schedule incoming [%d]" % len(to_schedule))`

`priority` should be considered here - sum of lengths of pools per priority

**Review by mtitov — COMMENTED** [review]
several small comments

File: src/radical/pilot/agent/scheduler/base.py [hunk 2]
```diff
@@ -410,9 +414,11 @@ def control_cb(self, topic, msg):
...
             with self._lock:
                 for uid in uids:
-                    if uid in self._waitpool:
-                        to_cancel.append(self._waitpool[uid])
-                        del self._waitpool[uid]
+                    for priority in self._waitpool:
+                        if uid in self._waitpool[priority]:
+                            task = self._waitpool[priority][uid]
+                            to_cancel.append(task)
+                            del self._waitpool[priority][uid]
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/agent/scheduler/base.py:421]: Thanks, Done

File: src/radical/pilot/agent/scheduler/base.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/scheduler/base.py:907]: thanks, done.

File: src/radical/pilot/task_description.py [hunk 3]
```diff
@@ -305,6 +306,12 @@ class TaskDescription(ru.TypedDict):
             the tasks did not manage to produce the expected output files to
             stage. Default False.
 
+        priority: (int, optional): The priority of the task.  Tasks with higher
+            priority will be scheduled first.  The default priority is 0.
+            The task process will not be strictly enforced strictly - under
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/task_description.py:311]: Both done, thanks.

**Review by mtitov — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**